fix(ProcessResponseBody): Avoids multiple calls, discussion about OnHttpStreamDone call

M4tteoP opened the pull request:
This PR aims to:
1. Address multiple calls of `ProcessResponseBody` happening when `SecResponseBodyAccess` is `off`. `OnHttpResponseBody` might be called multiple times, the call of `ProcessResponseBody` if the response body is not accessible was not followed by enabling the `processedResponseBody` boolean (and anticipated by the relative check).
   - Most likely it fixes https://github.com/corazawaf/coraza-proxy-wasm/issues/181#issuecomment-1540009474.
2. Discuss the usefulness of calling `ProcessResponseBody` during `OnHttpStreamDone()`. The concern has been raised here by @jcchavezs: https://github.com/corazawaf/coraza-proxy-wasm/pull/183#issuecomment-1528651275
   - During OnHttpStreamDone we can not enforce anymore any action, the interruption is indeed not even checked
   - I might only see some value in terms of triggering rules that might generate variables that will be logged, but it seems to be quite in conflict with the idea of optimizing the process and avoiding processing rules if we already know that we have to enforce a disruptive action (or that we can not do anything more)
   - The anomaly score would be further increased (if anything is found in the body), but again, it is a value that can just lead to the logs, being too late to enforce any action. It might even be quite misleading to log an anomaly score that has been computed after the point of actions enforcement
   - It has been introduced with https://github.com/corazawaf/coraza-proxy-wasm/commit/99466ec646c450949d8f335c12cc5b8f12c3338e, @anuraaga do you recall the reasoning that has been done that led to this call? 🙏. I added the remotion as a separate commit to easily revert it, the main point is to open a discussion around it :)

jcchavezs commented on wasmplugin/plugin.go:
```diff
@@ -592,21 +593,6 @@ func (ctx *httpContext) OnHttpStreamDone() {
 	tx := ctx.tx
 
 	if tx != nil {
-		if !tx.IsRuleEngineOff() && !ctx.interruptedAt.isInterrupted() {
```
Ping @anuraaga

anuraaga replied:
I think I did it just to be symmetric with processrequest. While interruptions can't be generated here, running in detectiononly mode or similar with phase 4 rules that apply to bodyless responses could still need it? Maybe not as much if multiphase can handle the rules, but it won't be able to handle all rules I guess

jcchavezs commented:
Yeah I think what @anuraaga says makes sense. Most likely we just need to
take this case in consideration.

On Tue, 16 May 2023, 04:45 Anuraag Agrawal, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In wasmplugin/plugin.go
> <https://github.com/corazawaf/coraza-proxy-wasm/pull/193#discussion_r1194542779>
> :
>
> > @@ -592,21 +593,6 @@ func (ctx *httpContext) OnHttpStreamDone() {
>  	tx := ctx.tx
>
>  	if tx != nil {
> -		if !tx.IsRuleEngineOff() && !ctx.interruptedAt.isInterrupted() {
>
> I think I did it just to be symmetric with processrequest. While
> interruptions can't be generated here, running in detectiononly mode or
> similar with phase 4 rules that apply to bodyless responses could still
> need it? Maybe not as much if multiphase can handle the rules, but it won't
> be able to handle all rules I guess
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/corazawaf/coraza-proxy-wasm/pull/193#discussion_r1194542779>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXOYARMKNIRKYRGFJSHCBLXGLS4LANCNFSM6AAAAAAX72IP4Y>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

M4tteoP commented:
Thanks Rag! I reverted the change and added a Warning when ProcessResponseBody is called during OnHttpStreamDone. My main concern is about misleading logs and anomaly scores generated.
I see running in `detectionOnly` mode mostly as a way to first look at the WAF behavior before turning it on. A user would see triggered rules and increased anomaly scores thinking that they will be the rules/scores effective as soon as the WAF is turned on, but it is not the case. I don't know if the Warning is enough, but It should be a first step to raise user's attention when reading the logs.

jcchavezs commented on wasmplugin/plugin.go:
```diff
@@ -597,6 +598,7 @@ func (ctx *httpContext) OnHttpStreamDone() {
 			// phase that still need to be executed. If they haven't been executed yet, and there has not been a previous
 			// interruption, now is the time.
 			if !ctx.processedResponseBody {
+				ctx.logger.Warn().Msg("Running ProcessResponseBody in OnHttpStreamDone, triggered actions will not be enforced from now on")
```
Since this is an expected operation we can use `Info` log level or debug. Also could you ellaborate more and tell that it will trigger another log from coraza (if that is the case?)

M4tteoP replied:
I went for the `Warn` log level because I was not happy about a case like the following:
Request:
```
curl -I 'http://localhost:8080/status/406'
```
Rule:
```
"SecRuleEngine On",
"SecRule REQUEST_URI \"@rx /status\" \"id:101,phase:4,log,t:lowercase,deny\""
```
Logs:
```
wasm log coraza-filter my_vm_id: [client "172.19.0.1"] Coraza: Warning.  [file ""] [line "986"] [id "101"] [rev ""] [msg ""] [data ""] [severity "emergency"] [ver ""] [maturity "0"] [accuracy "0"] [hostname "172.19.0.2"] [uri "/status/406"] [unique_id "RPLGaNMdNlLPGDUBjpd"]
wasm log coraza-filter my_vm_id: Finished tx_id="cDBqsDMNnWgNYyZSjrp" context_id=3
```

It seems like the rule worked properly, but the disruptive action was not enforced in the end.


Moved to `Info`.
